### PR TITLE
Streaming Position Controller

### PR DESCRIPTION
**Pull request by hello-fazil**
### Streaming Position Control
This PR adds a streaming position control interface where the goal joint pose array can published to the the topic `/joint_pose_cmd` as `Float64MultiArray` message type. 
- This feature can be used in `position` and `navigation` modes. 
- Streaming position control can be activated and deactivated by calling the services `/activate_streaming_position` and `/deactivate_streaming_position`. The current activation state is published in topic `/is_streaming_position`.
#### Framing a Goal Pose Array
- The goal pose array should be of a fixed length that is determined by the robot tool config.
- Indexing helper methods are provided to frame the pose array.
- Null value goals must be given as absolute Zero.
- `BASE_TRANSLATE` and `BASE_ROTATE` are **incremental** poses.
- Only either one of `BASE_TRANSLATE` or `BASE_ROTATE` must hold a non-zero value. Giving non-zero value for both the joints will be dropped for base.
- Base position commands will be dropped in Navigation mode.

```python
from hello_helpers.joint_qpos_conversion import get_Idx
pose = np.zeros(self.Idx.num_joints)
Idx = get_Idx('eoa_wrist_dw3_tool_sg3')
pose[Idx.LIFT] = 0.6
pose[Idx.Idx.ARM] = 0.2
pose[Idx.GRIPPER] = 100
pose[Idx.WRIST_ROLL] = 0.1
pose[Idx.WRIST_PITCH] = -np.pi/4
pose[Idx.WRIST_YAW] = 0.0
pose[Idx.HEAD_PAN] = 0.4
pose[Idx.HEAD_TILT] = 0.3
pose[Idx.BASE_TRANSLATE] = 0.4
pose[Idx.BASE_ROTATE] = 0.0
```

### Comments
- Currently, this feature implementation uses a [std_msgs/Float64MultiArray](https://docs.ros.org/en/melodic/api/std_msgs/html/msg/Float64MultiArray.html), which requires careful indexing and complications around writing code. I think using the [sensor_msg/JointState](https://docs.ros.org/en/noetic/api/sensor_msgs/html/msg/JointState.html)  might be a better solution as provides the following benefits:
   - Removes complicated Indexing by providing an order list of joint names str
   - Provides a way to set Velocity, acceleration, and effort values with a streaming position controller.

Here is an example test client node. This script is also wrapped up as an test script:
```python
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float64MultiArray
from hello_helpers.joint_qpos_conversion import get_Idx, JointStateMapping
from sensor_msgs.msg import JointState
from std_srvs.srv import Trigger
from hello_helpers.gripper_conversion import GripperConversion
from rclpy.callback_groups import ReentrantCallbackGroup
import numpy as np
import time



class JointPosePublisher(Node):
    def __init__(self):
        rclpy.init()
        super().__init__('float_array_publisher')
        self.reentrant_cb = ReentrantCallbackGroup()
        self.publisher_ = self.create_publisher(Float64MultiArray, 'joint_pose_cmd', 10,callback_group=self.reentrant_cb)
        # subscribe to joint states
        self.joint_state = JointState()
        self.joint_states_subscriber = self.create_subscription(JointState, 
                                                                '/stretch/joint_states', 
                                                                self.joint_states_callback, 10,callback_group=self.reentrant_cb)
        self.Idx = get_Idx('eoa_wrist_dw3_tool_sg3')
        self.switch_to_position_mode_service = self.create_client(Trigger, '/switch_to_position_mode',callback_group=self.reentrant_cb)

        self.switch_to_navigation_mode_service = self.create_client(Trigger, '/switch_to_navigation_mode',callback_group=self.reentrant_cb)

        self.activate_streaming_position_service = self.create_client(Trigger, '/activate_streaming_position', callback_group=self.reentrant_cb)
        self.deactivate_streaming_position_service = self.create_client(Trigger, '/deactivate_streaming_position', callback_group=self.reentrant_cb)

        while not self.switch_to_position_mode_service.wait_for_service(timeout_sec=2.0):
            self.get_logger().info("Waiting on '/switch_to_position_mode' service...")
        self.gripper_conversion = GripperConversion()

    def joint_states_callback(self, msg):
        self.joint_state = msg

    def switch_to_position_mode(self):
        trigger_request = Trigger.Request()
        trigger_result = self.switch_to_position_mode_service.call_async(trigger_request)
        return

    def activate_streaming_position(self):
        trigger_request = Trigger.Request()
        trigger_result = self.activate_streaming_position_service.call_async(trigger_request)
        time.sleep(1)

    def deactivate_streaming_position(self):
        trigger_request = Trigger.Request()
        trigger_result = self.deactivate_streaming_position_service.call_async(trigger_request)
        return


    def switch_to_navigation_mode(self):
        trigger_request = Trigger.Request()
        trigger_result = self.switch_to_navigation_mode_service.call_async(trigger_request)
        return

    def get_joint_status(self):
        j_status =  self.parse_joint_state(self.joint_state)
        pose = np.zeros(self.Idx.num_joints)
        pose[self.Idx.LIFT] = j_status[JointStateMapping.ROS_LIFT_JOINT]
        pose[self.Idx.ARM] = sum(j_status[joint] for joint in JointStateMapping.ROS_ARM_JOINTS)
        pose[self.Idx.GRIPPER] = j_status[JointStateMapping.ROS_GRIPPER_FINGER]
        pose[self.Idx.WRIST_ROLL] = j_status[JointStateMapping.ROS_WRIST_ROLL]
        pose[self.Idx.WRIST_PITCH] = j_status[JointStateMapping.ROS_WRIST_PITCH]
        pose[self.Idx.WRIST_YAW] = j_status[JointStateMapping.ROS_WRIST_YAW]
        pose[self.Idx.HEAD_PAN] = j_status[JointStateMapping.ROS_HEAD_PAN]
        pose[self.Idx.HEAD_TILT] = j_status[JointStateMapping.ROS_HEAD_TILT]
        return pose

    def parse_joint_state(self, joint_state_msg):
        joint_status = {}
        for name, position in zip(joint_state_msg.name, joint_state_msg.position):
            joint_status[name] = position
        return joint_status

    def publish_joint_pose(self, joint_pose):
        msg = Float64MultiArray()
        msg.data = list(joint_pose)
        self.publisher_.publish(msg)
        self.get_logger().info('Publishing: "%s"' % msg.data)

    def wait_until_at_setpoint(self, goal_qpos):
        while abs(goal_qpos[:-2] - joint_pose_publisher.get_joint_status()[:-2]).mean() > 0.01:
            rclpy.spin_once(self)
            time.sleep(0.01)

if __name__ == '__main__':
    joint_pose_publisher = JointPosePublisher()
    rclpy.spin_once(joint_pose_publisher)

    Idx = get_Idx('eoa_wrist_dw3_tool_sg3')

    # joint_pose_publisher.switch_to_navigation_mode()
    joint_pose_publisher.activate_streaming_position()

    qpos = np.zeros(Idx.num_joints)
    qpos[Idx.LIFT] = 0.6
    qpos[Idx.ARM] = 0
    qpos[Idx.WRIST_PITCH] = 0
    qpos[Idx.WRIST_ROLL] = 0
    qpos[Idx.WRIST_YAW] = 0
    qpos[Idx.GRIPPER] = joint_pose_publisher.gripper_conversion.robotis_to_finger(0)
    qpos[Idx.BASE_TRANSLATE] = 0
    qpos[Idx.BASE_ROTATE] = 0
    qpos[Idx.HEAD_PAN] = 0
    qpos[Idx.HEAD_TILT] = 0
    joint_pose_publisher.publish_joint_pose(qpos)
    joint_pose_publisher.wait_until_at_setpoint(qpos)

    i = 0
    while i<100:
        i = 1 + i
        rclpy.spin_once(joint_pose_publisher)
        qpos = joint_pose_publisher.get_joint_status()
        qpos[Idx.LIFT] = qpos[Idx.LIFT] + 0.05
        qpos[Idx.ARM] = qpos[Idx.ARM] + 0.05
        qpos[Idx.WRIST_PITCH] = qpos[Idx.WRIST_PITCH] + 0.1
        qpos[Idx.WRIST_ROLL] = qpos[Idx.WRIST_ROLL] + 0.1
        qpos[Idx.WRIST_YAW] = qpos[Idx.WRIST_YAW] + 0.1
        qpos[Idx.HEAD_PAN] = qpos[Idx.HEAD_PAN] + 0.1
        qpos[Idx.HEAD_TILT] = qpos[Idx.HEAD_TILT] + 0.1
        qpos[Idx.GRIPPER] = qpos[Idx.GRIPPER] + 0.1
        qpos[Idx.BASE_TRANSLATE] = 0.01
        qpos[Idx.BASE_ROTATE] = 0.0
        joint_pose_publisher.publish_joint_pose(qpos)
        time.sleep(1/15)
    i = 0
    while i<100:
        i = 1 + i
        rclpy.spin_once(joint_pose_publisher)
        qpos = joint_pose_publisher.get_joint_status()
        qpos[Idx.LIFT] = qpos[Idx.LIFT] - 0.05
        qpos[Idx.ARM] = qpos[Idx.ARM] - 0.05
        qpos[Idx.WRIST_PITCH] = qpos[Idx.WRIST_PITCH] - 0.1
        qpos[Idx.WRIST_ROLL] = qpos[Idx.WRIST_ROLL] - 0.1
        qpos[Idx.WRIST_YAW] = qpos[Idx.WRIST_YAW] - 0.1
        qpos[Idx.HEAD_PAN] = qpos[Idx.HEAD_PAN] - 0.1
        qpos[Idx.HEAD_TILT] = qpos[Idx.HEAD_TILT] - 0.1
        qpos[Idx.GRIPPER] = qpos[Idx.GRIPPER] - 0.1
        qpos[Idx.BASE_ROTATE] = -0.05
        qpos[Idx.BASE_TRANSLATE] = 0.0
        joint_pose_publisher.publish_joint_pose(qpos)
        time.sleep(1/15)

    joint_pose_publisher.destroy_node()
    rclpy.shutdown()
```

### CODE REVIEW

File: stretch_description/urdf/stretch_base_imu.xacro [hunk 1]
```diff
@@ -8,14 +8,14 @@
         xyz="0.00300349280517617 0.00149777182047641 -0.00193103885249443"
         rpy="0 0 0" />
       <mass
-        value="0.000901473198307758" />
+        value="0.00092" />
       <inertia
```
Comment (hello-cpaxton) [inline stretch_description/urdf/stretch_base_imu.xacro]: Why was this changed? what is it here for?

File: stretch_core/stretch_core/stretch_driver.py [hunk 2]
```diff
@@ -108,6 +112,62 @@ def set_mobile_base_velocity_callback(self, twist):
...
+        if self.get_clock().now() - self.streaming_controller_lt.last_update_time> 5:
+            if STREAMING_POSITION_DEBUG:
+                print('Reset Streaming position looptimer after 5s no message received.')
+            self.streaming_controller_lt.reset()
+        qpos = msg.data
+        self.move_to_position(qpos)
+        self.robot_mode_rwlock.release_read()
+        self.streaming_controller_lt.update()
+    
+    def move_to_position(self, qpos):
```
Comment (hello-cpaxton) [inline stretch_core/stretch_core/stretch_driver.py:141]: I know we don't do that in this repo but I still think it would be good to move towards type hint + docstring for things like this

what dimensiionality should qpos be for example

File: stretch_core/stretch_core/stretch_driver.py [hunk 3]
```diff
@@ -108,6 +112,62 @@ def set_mobile_base_velocity_callback(self, twist):
...
+            if 'wrist_pitch' in self.robot.end_of_arm.joints:
+                self.robot.end_of_arm.move_to('wrist_pitch', qpos[Idx.WRIST_PITCH])
+            if 'wrist_roll' in self.robot.end_of_arm.joints:
+                self.robot.end_of_arm.move_to('wrist_roll', qpos[Idx.WRIST_ROLL])
+            self.robot.head.move_to('head_pan', qpos[Idx.HEAD_PAN])
+            self.robot.head.move_to('head_tilt', qpos[Idx.HEAD_TILT])
+            if abs(qpos[Idx.BASE_TRANSLATE]) > 0.0 and abs(qpos[Idx.BASE_ROTATE]) > 0.0 and self.robot_mode != 'position':
+                self.get_logger().error('Cannot move base in both translation and rotation at the same time in position mode')
+            elif abs(qpos[Idx.BASE_TRANSLATE]) > 0.0 and self.robot_mode == 'position':
+                self.robot.base.translate_by(qpos[Idx.BASE_TRANSLATE])
```
Comment (hello-cpaxton) [inline stretch_core/stretch_core/stretch_driver.py:162]: As mentioned on Slack I dont love using Idx in this way

are dictionaries with strings (ie JointState) still fast enough?

File: hello_helpers/src/hello_helpers/hello_misc.py [hunk 4]
```diff
@@ -626,3 +627,35 @@ def preprocess_gripper_trajectory(trajectory):
     gripper_index = trajectory.joint_names.index(present_gripper_joints[0])
     trajectory.joint_names[gripper_index] = 'stretch_gripper'
     return trajectory
+
+
+class LoopTimer:
```
Comment (hello-cpaxton) [inline hello_helpers/src/hello_helpers/hello_misc.py:632]: This class feels like it's been added to our libraries half a dozen times

maybe hello_helpers should be its own pip install able package for things like this

File: hello_helpers/src/hello_helpers/joint_qpos_conversion.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python3
+
+class UnsupportedToolError(Exception):
+    pass
+
+def get_Idx(tool_name):
+    """
+    Returns the Idx class corresponding to the tool name.
+    """
+    if tool_name == 'eoa_wrist_dw3_tool_sg3':
```
Comment (hello-cpaxton) [inline hello_helpers/src/hello_helpers/joint_qpos_conversion.py:10]: This is a nice system for handling this imo

File: stretch_core/test/test_streaming_position_commands.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import pytest
+import rclpy
+
+from common.launch_descriptions import stretch_driver_ld
+from common.client_nodes.joint_pose_streaming import JointPosePublisher
+import numpy as np
+import time
+
+@pytest.mark.launch(fixture=stretch_driver_ld)
+def test_streaming_position_commands():
```
Comment (hello-binit) [inline stretch_core/test/test_streaming_position_commands.py:11]: Nice test!

File: stretch_description/urdf/stretch_base_imu.xacro (hunk 1 above)
Comment (hello-binit) [reply stretch_description/urdf/stretch_base_imu.xacro]: Agreed, this should be reverted.

File: stretch_core/stretch_core/stretch_driver.py [hunk 7]
```diff
@@ -108,6 +112,62 @@ def set_mobile_base_velocity_callback(self, twist):
...
+
+        
+        if self.get_clock().now() - self.streaming_controller_lt.last_update_time> 5:
+            if STREAMING_POSITION_DEBUG:
+                print('Reset Streaming position looptimer after 5s no message received.')
+            self.streaming_controller_lt.reset()
+        qpos = msg.data
+        self.move_to_position(qpos)
+        self.robot_mode_rwlock.release_read()
+        self.streaming_controller_lt.update()
```
Comment (hello-binit) [inline stretch_core/stretch_core/stretch_driver.py]: Do we want the overhead of this profiler running constantly? Perhaps it should be default off, and able to be turned on with a flag. I'd also love to see a performance test that monitors this timer.

**Review by hello-binit — CHANGES_REQUESTED** [review]
Hey @hello-fazil, nice work! How does this streaming interface interact with the FJT action server? It sounds like both interfaces can be active at the same time. Since this can become complicated, I'd argue there should be a `streaming`/`streaming_nav` mode(s) where the FJT is turned off.

Can you add a few notes about performance? What rate of control do you expect to achieve for the whole body? Have you profiled with streaming + Twist commands?

File: stretch_description/urdf/stretch_base_imu.xacro (hunk 1 above)
Comment (hello-fazil) [reply stretch_description/urdf/stretch_base_imu.xacro]: Actually right value is the changes which this PR brings which will be match with what we provide from the Stretch URDF package.
https://github.com/hello-robot/stretch_urdf/blob/main/stretch_urdf/SE3/xacro/stretch_base_imu.xacro#L11:L18
Therefore I think it is okay to leave this change.

**Comment by hello-fazil** [comment]
@hello-binit Currently FJT invalidates all the goals received when the Streaming position is activated.
https://github.com/hello-robot/stretch_ros2/blob/2fa56f4227efcc79878491c23880317815a548c9/stretch_core/stretch_core/joint_trajectory_server.py#L121:L122

**Comment by hello-fazil** [comment]
@hello-binit I like the idea of adding two modes `streaming` and `streaming_navigation` instead of activating and deactivating a streaming controller flag shared between the current position and navigation mode. But the only issue I see is in increasing the number of control modes causing confusions or may be this already exists. 
I will be doing a subsequent PR on top of this which will use `JointState` sensor message instead of `Float64MultiArray`. May be I can address this in that PR and discuss further.

**Comment by hello-fazil** [comment]
@hello-binit In terms of performance I get an average ~26Hz streaming control execution when published at 30Hz from the client within the robot. Publishing `/cmd_vel` does not seem to affect the above rate (Fyi, this was tested with the bumped 5 thread executor)

**Comment by hello-fazil** [comment]
@hello-binit  This PR has been functionally well tested by Stretch AI team @hello-cpaxton , therefore I am merging this PR in as it is critical for them. Additional concerns on having a `streaming_position` and `streaming_nav` mode can be discussed in the subsequent PR that will introduce using `JointState` message receiving goals.
